csharp/extractor/Semmle.Extraction.CSharp/Entities/PreprocessorDirectives/IfDirective.cs

```diff
@@ -0,0 +1,40 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Collections.Generic;
+using System.IO;
+
+namespace Semmle.Extraction.CSharp.Entities
+{
+    internal class IfDirective : PreprocessorDirective<IfDirectiveTriviaSyntax>, IExpressionParentEntity
+    {
+        private readonly List<IIfSiblingDirective> branches = new List<IIfSiblingDirective>();
+
+        public IfDirective(Context cx, IfDirectiveTriviaSyntax trivia)
+            : base(cx, trivia)
+        {
+        }
+
+        public bool IsTopLevelParent => true;
+
+        protected override void PopulatePreprocessor(TextWriter trapFile)
+        {
+            trapFile.directive_ifs(this, trivia.BranchTaken, trivia.ConditionValue);
+
+            Expression.Create(cx, trivia.Condition, this, 0);
+        }
+
+        internal void Add(IIfSiblingDirective branch)
+        {
+            branches.Add(branch);
+        }
+
+        internal void WriteBranches(EndIfDirective endif)
+        {
+            cx.TrapWriter.Writer.directive_if_endif(this, endif);
+            var siblings = 0;
+            foreach (var branch in branches)
+            {
+                cx.TrapWriter.Writer.directive_if_siblings(this, branch, siblings++);
+            }
+        }
```

Comment:
I would prefer if instead the `EndIfDirective` constructor had a `IfDirective parent` parameter, and that the `ElifDirective` and `ElseDirective` constructors had `IfDirective parent, int child` parameters, and we did this population there instead. That would mean keeping track of an extra integer for the child index in the stack in `DirectiveVisitor.cs` though.

Reply:
Then we can actually get rid of the `directive_if_endif` and `directive_if_siblings` relations, as the information can be put into `directive_endifs` and `directive_elifs`/`directive_elses` instead.